packages/flutter_tools/lib/src/base/os.dart

```diff
@@ -463,17 +459,40 @@ class _WindowsUtils extends OperatingSystemUtils {
 /// directory or the current working directory if none specified.
 /// Return null if the project root could not be found
 /// or if the project root is the flutter repository root.
-String findProjectRoot([ String directory ]) {
+String? findProjectRoot(FileSystem fileSystem, [ String? directory ]) {
   const String kProjectRootSentinel = 'pubspec.yaml';
-  directory ??= globals.fs.currentDirectory.path;
+  directory ??= fileSystem.currentDirectory.path;
```

Comment:
I'm really surprised this didn't promote 🤔

Reply:
I think if it was setting to a local it would, but because it's passed as a `String?` parameter, when I don't add the `!`s I get:
```
error: The argument type 'String?' can't be assigned to the parameter type 'String'. (argument_type_not_assignable at [flutter_tools] lib/src/base/os.dart:469)
```
Is that what you meant?

Reply:
I think `??=` normally promotes though, and parameters can be promoted unlike fields. Maybe this is more of a language question. For example:

```
String? foo;
foo ??= 'bar'
if (foo != null) { // Err: foo cannot be null?

}
```

Reply:
Jonah figured out this is because it's being reset in the `while` loop.  When I get rid of `while` it complains about `!`, when I get rid of `directory = parent` in the loop it also promotes.  Even though `parent` is a `String`.